### PR TITLE
METRON-1634: Alerts UI add comment doesn't immediately show up.

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This PR changes the way updates to alert comments work in the Alerts UI.  Previously when adding/removing a comment, the latest version of an alert is retrieved after a 200 status code is received.  However there is latency involved when updating an alert so it's possible the old version is still displayed.

This PR treats updates to alert comments optimistically.  The change is applied immediately in the UI and reverted only if an error is returned on update.  A lookup of the alert is no longer needed.  

Changes include:
- adding logic to immediately add a comment in the client side code
- a slight refactor of the setComments function to make it more flexible
- adding an error callback property to the patchAlert function
- implementing error callback functions for adding/removing comments to revert to previous state

In both cases of adding/removing comments, reverting state follows the pattern of copying the previous list of comments and falling back to the copy on failure.  I used the `slice` method to do this.  Would be interested to hear if this is the right approach for copying arrays.

This has been tested in full dev.  Navigate to the Alerts UI and open the details panel for an alert.  Select the comments tab and add a comment.  The change should be displayed immediately.  Removing a comment should also be reflected immediately.  To test the case of an error being returned from the back end, I stopped the REST application and then tried to add a comment.  You should see the alert added for no more than a split second before reverting to the previous state.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by merrimanr** [comment]
Latest commit should address your comments.  Let me know what you think.

File: metron-interface/metron-alerts/src/app/alerts/alert-details/alert-details.component.ts [hunk 1]
```diff
@@ -223,21 +222,23 @@ export class AlertDetailsComponent implements OnInit {
 
   onAddComment() {
     let alertComment = new AlertComment(this.alertCommentStr, this.authenticationService.getCurrentUserName(), new Date().getTime());
```
Comment (tiborm) [inline metron-interface/metron-alerts/src/app/alerts/alert-details/alert-details.component.ts]: I think this is the newComment...

File: metron-interface/metron-alerts/src/app/alerts/alert-details/alert-details.component.ts [hunk 2]
```diff
@@ -223,21 +222,23 @@ export class AlertDetailsComponent implements OnInit {
...
-    let tAlertComments = this.alertCommentsWrapper.map(alertsWrapper => alertsWrapper.alertComment);
-    tAlertComments.unshift(alertComment);
-    this.patchAlert(new Patch('add', '/comments', tAlertComments));
+    let newComments = this.alertCommentsWrapper.map(alertsWrapper => alertsWrapper.alertComment);
```
Comment (tiborm) [inline metron-interface/metron-alerts/src/app/alerts/alert-details/alert-details.component.ts]: and this is the alertComments extracted from the commentWrappers.

**Comment by merrimanr** [comment]
Ok latest commit should do it.

**Comment by merrimanr** [comment]
@justinleet I changed the logic to reverse the operation rather than reverting back to a copy.  This means that a comment will be removed from the list on a failed add and a comment will be added back to the list on a failed remove.  The list is sorted by timestamp after a comment is put back to preserve the order.  This should make different add/remove operations independent of each other.  Let me know what you think.

**Comment by tiborm** [comment]
+1 nice improvement

**Pull request closed without merging**